src/Module/System/Update.php

```diff
@@ -4379,4 +4382,112 @@ public static function update_540002(): bool
 
         return $retval;
     }
+
+    /** update_54003
+     *
+     * Add new tables and modify others to support catalog access feature
+     *
+     */
+    public static function update_540003(): bool
+    {
+        $retval     = true;
+        $collation  = (AmpConfig::get('database_collation', 'utf8mb4_unicode_ci'));
+        $charset    = (AmpConfig::get('database_charset', 'utf8mb4'));
+        $engine     = ($charset == 'utf8mb4') ? 'InnoDB' : 'MYISAM';
+
+        // Add the new catalog_access_group table
+        $sql = "CREATE TABLE IF NOT EXISTS `catalog_access_group` (`id` int(11) UNSIGNED NOT NULL AUTO_INCREMENT,`name` varchar(128) COLLATE utf8mb4_unicode_ci DEFAULT NULL,PRIMARY KEY (`id`),UNIQUE KEY `name` (`name`)) ENGINE=$engine DEFAULT CHARSET=$charset COLLATE=$collation;";
+        $retval &= (Dba::write($sql) !== false);
+
+        // Add the new default_catalog_access_group table to prevent deletion of the default group
+        $sql = "CREATE TABLE `default_catalog_access_group` (`default_catalog_access_group_id` int(11) UNSIGNED PRIMARY KEY,FOREIGN KEY (default_catalog_access_group_id) references catalog_access_group (id));";
+        $retval &= (Dba::write($sql) !== false);
+
+        // Add the new catalog_access table
+        $sql = "CREATE TABLE IF NOT EXISTS `catalog_access` (`access_group_id` int(11) UNSIGNED NOT NULL,`catalog_id` int(11) UNSIGNED NOT NULL,`enabled` tinyint(1) UNSIGNED NOT NULL DEFAULT 0,FOREIGN KEY (`access_group_id`) REFERENCES catalog_access_group(`id`) ON DELETE CASCADE, FOREIGN KEY (`catalog_id`) REFERENCES catalog(`id`) ON DELETE CASCADE) ENGINE=$engine DEFAULT CHARSET=$charset COLLATE=$collation;";
```

Comment:
errors making that table. i deleted the second update so maybe that's incorrect compared to the first db update?

```
Error_query SQL: CREATE TABLE IF NOT EXISTS `catalog_access` (`access_group_id` int(11) UNSIGNED NOT NULL, `catalog_id` int(11) UNSIGNED NOT NULL, `enabled` tinyint(1) UNSIGNED NOT NULL DEFAULT 0,FOREIGN KEY (`access_group_id`) REFERENCES catalog_access_group(`id`) ON DELETE CASCADE, FOREIGN KEY (`catalog_id`) REFERENCES catalog(`id`) ON DELETE CASCADE) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```

```
Error_query MSG: ["HY000",1005,"Can't create table `ampachetest`.`catalog_access` (errno: 150 \"Foreign key constraint is incorrectly formed\")"]
```

Reply:
I'm not seeing this problem.  I copied the code and used it directly in mysql:
```
MariaDB [(none)]> use ampache
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [ampache]> CREATE TABLE IF NOT EXISTS `catalog_access_group` (`id` int(11) UNSIGNED NOT NULL AUTO_INCREMENT,`name` varchar(128) COLLATE utf8mb4_unicode_ci DEFAULT NULL,PRIMARY KEY (`id`),UNIQUE KEY `name` (`name`)) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
Query OK, 0 rows affected (0.029 sec)

MariaDB [ampache]> CREATE TABLE `default_catalog_access_group` (`default_catalog_access_group_id` int(11) UNSIGNED PRIMARY KEY,FOREIGN KEY (default_catalog_access_group_id) references catalog_access_group (id));
Query OK, 0 rows affected (0.046 sec)

MariaDB [ampache]> CREATE TABLE IF NOT EXISTS `catalog_access` (`access_group_id` int(11) UNSIGNED NOT NULL, `catalog_id` int(11) UNSIGNED NOT NULL, `enabled` tinyint(1) UNSIGNED NOT NULL DEFAULT 0,FOREIGN KEY (`access_group_id`) REFERENCES catalog_access_group(`id`) ON DELETE CASCADE, FOREIGN KEY (`catalog_id`) REFERENCES catalog(`id`) ON DELETE CASCADE) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
Query OK, 0 rows affected (0.030 sec)
```
The tables have to be created in that order or else the KEY references don't work.  I'm going to restore the database and try it from the code,.

Reply:
ok, that's good, it could also be from me and my database. i'll do some more tesitng tomorrow but at least we're on our way!